File: p070e/climb_stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        a, b = 1, 1
        for _ in range(1, n):
            a, b = b, a + b
        return b


# TESTS
for n, expected in [
    (2, 2),
    (3, 3),
    (20, 10946),
]:
    sol = Solution()
    actual = sol.climbStairs(n)
    print("The ways to climb", n, "stairs ->", actual)
    assert actual == expected

File: p070e/climb_stairs.js
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function (n) {
  let dp = Array.from({ length: n + 1 }, (v, k) => 1);
  for (let i = 2; i <= n; ++i) {
    dp[i] = dp[i - 1] + dp[i - 2];
  }
  return dp[n];
};
var climbStairsFab = function (n) {
  let [a, b] = [1, 1];
  while (--n) {
    [a, b] = [b, a + b];
  }
  return b;
};
// TEST
for (let n = 1; n <= 45; ++n) {
  console.log(
    'There are',
    climbStairsFab(n),
    'ways to climb to the top of',
    n,
    'steps'
  );
  console.assert(climbStairs(n) === climbStairsFab(n));
}
